RSE manager: handle RSEChecksumUnavailabale depending on RSE settings Fix #1149

berghaus opened the pull request:
RSE manager: handle RSEChecksumUnavailabale depending on RSE settings Fix #1149

Commit:
RSE Manager: failure to retrieve checksum from RSE raises RSEChecksumUnavailabale

Commit:
RSE manager: handle RSEChecksumUnavailabale depending on RSE settings

If RSE is flagged not to support checksums assume data is valid, else
assume there is a problem with the RSE.

berghaus commented on lib/rucio/rse/rsemanager.py:
```diff
@@ -317,7 +317,7 @@ def exists(rse_settings, files):
     for f in files:
         exists = None
-        if (type(f) is str) or (type(f) is unicode):
+        if isinstance(f, str) or isinstance(f, type(u'')):
```
Flake8 complained here, so I added this. seems OK. Not pretyt with the `type(u'')` though ...

mlassnig replied:
isinstance("asdfasdf", basestring)

tests for either string or unicode

TWAtGH replied:
I think basestring doesn't exist in python 3

mlassnig replied:
yep, in py3 string doesn't exist anymore ;-)

for py3 we only do isinstance(..., unicode)

berghaus replied:
In py3 only str exists, in py2 str and unicode, hence (one of the) falke8 complaints. How about:

```python
if sys.version_info.major == 3:
    string_types = (str,)
else:
    string_types = (str, unicode)
if isinstance(f, string_types):
```

berghaus replied:
Mind you flake still complains about this: `F821 undefined name 'unicode'`

berghaus replied:
How about:
```python
try:
    unicode
except NameError:
    unicode = str
```

berghaus replied:
So this is giving me a bad of a bad smell: I need to do this in each source file where we use the `unicode` keyword ... but I don't quite know how to do this as a module.

berghaus replied:
So what I could do is make a `common/python3workaround.py` containing:
```python
try:
    unicode
except NameError:
    unicode = str
```
then in any file which uses `unicode`
```python
try:
    from rucio.common.python3workaround import unicode
except ImportError:
    pass
```
Long live exceptions?!

vingar replied:
we did something like that in some part of the code:
```
try:
    AUTHORIZED_VALUE_TYPES = (float, int, unicode)
except NameError:
    AUTHORIZED_VALUE_TYPES = (float, int, str)
```
generating the possible type values and then check against them

TWAtGH replied:
```
if sys.version_info[0] == 2:
	strclass = basestring
elif sys.version_info[0] == 3:
	strclass = str
else:
	raise
```
Wondering why people in Python always try to replaces if's with exceptions.

vingar replied:
it's actually the recommended way. I read several times this advice.

TWAtGH replied:
Only because a lot of people do it or it's marked as solution on stackoverflow it doesn't need to be recommended :D

vingar replied:
... mentioned with some motivation about the performance ;) but I doubt I can convince you anyway ;)

berghaus replied:
@TWAtGH this is python [EAFP](https://docs.python.org/3/glossary.html#term-eafp) style.

TWAtGH replied:
Yes I know but this code will only be executed once per start of a Rucio process. So this is no place were we should prefer performance over readability. Also EAFP should only be used when you know that a specific case is very exceptional and you don't know this in our case do you?

berghaus replied:
At the moment, the rare case is running in python3. Plus the exception avoids adding the sys module.

TWAtGH replied:
Not adding the sys module is a good point but I would guess that it's very unlikely that sys is not already loaded into memory, since it gets imported in some core areas of Rucio.
Again this is no point where performance matters because the impact is nearly nonexistent. Even if performance would matter the except cases should be really really rare.
The impact on the readability and maintainability would be huge though. One can immediately see that you're doing a differentiation between Python versions there. And think about adding more stuff that needs to be distinguished. With your approach you would end up adding a lot of try/excepts.

vingar replied:
Feature detection is already at various places of the code, following this:
https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection

I would prefer to be consistent 

About unicode:
https://docs.python.org/3/howto/pyporting.html#text-versus-binary-data

berghaus commented on lib/rucio/rse/rsemanager.py:
```diff
@@ -414,13 +414,17 @@ def upload(rse_settings, lfns, source_dir=None, force_pfn=None, force_scheme=Non
-                        valid = stats['filesize'] == lfn['filesize']
-                except NotImplementedError:
-                    valid = True if rse_settings['verify_checksum'] is False else False
+                    try:
```
The only other way I could think of doing this is an `isinstance` in the except clause, these nested try's seemed cleaner.

mlassnig commented on lib/rucio/common/exception.py:
```diff
@@ -849,3 +849,13 @@ def __init__(self, *args, **kwargs):
+    """
+    def __init__(self, *args, **kwargs):
+        super(RSEChecksumUnavailabale, self).__init__(*args, **kwargs)
+        self._message = "RSE Checksum unavailabale."
```
typo ;-)

mlassnig replied:
oh, just realised that the class itself has the same typo

berghaus replied:
Herp, time for me to go fix some typos.

Commit:
RSE manager: replace explicit references to unicode type with STRING_TYPES

Python 3 has changed the way strings (binary vs text) are handled. Part of this
change is the removal of the unicode type. To adress this in a consistent way
STRING_TYPES is constructed detecting which types are available.
